### ISSUE TITLE
ImageOutput: AppendSubimage creates broken subimages when using multiple channelformats

### ISSUE THREAD

**Issue by marcomeyerVFX** [OpenImageIO/oiio#1306]
I am trying to create multipart exrs from regular exrs using oiio's python bindings (to avoid using OpenEXR 2.0's "exrmultipart" tool, due to it's limited functionality), but appending subimages only creates healthy image data as long as the channelformat is the same for all subimages. Appending a subimage with FLOAT-only channels to a HALF-only subimage stack causes the FLOAT subimage to be unreadable, while all other subimages maintain healthy and accessible. Only workaround I have so far is to create FLOAT-only multipart exrs which is not really a workaround and not what i want, of course.
Can anyone confirm this?

OpenImageIO 1.6.8

**Comment by Agartner** [comment]
I'm able to do this with 1.5 at the moment though I think there may be an open pull request to fix a m_pixletype bug in the exroutput plugin which I have actually pulled down for my local install. What API calls are you using to set up/write the multi part output?

**Comment by marcomeyerVFX** [comment]
Here's a short example using files from the oiiotool-subimage testsuite to demonstrate the issue. subimageB1 has been converted to float for this:
 
```python
imgs = ["subimage1.exr", "subimage2.exr", "subimageB1_FLOAT.exr", "subimageD3.exr"]
out_file = "si_mixed_cf.exr"
new_specs = []
new_pixels = []
for i in imgs:
    image_in = oiio.ImageInput.open(i)
    spec = image_in.spec()
    pixels = image_in.read_image(oiio.FLOAT)
    new_specs.append(spec)
    new_pixels.append(pixels)
    image_in.close()

output = oiio.ImageOutput.create(out_file)
output.open(out_file, tuple(new_specs))

for i, s in enumerate(new_specs):
    if i > 0:
        output.open(out_file, s, oiio.AppendSubimage)
    output.write_image(new_pixels[i])
output.close()
```

**Comment by lgritz** [comment]
Thanks for the example, I'll take a look. At first glance, it appears that you're doing the right thing. Let me try to reproduce.

**Comment by Agartner** [comment]
Not sure if it makes a difference but I figure I'll note it anyway.  API call wise the only difference I have is I'm using write_scanlines rather than write_image for each subimage. Also for the first subimage I don't use oiio.AppendSubimage, just regular write_scanlines. Then all the subsequent ones are using the AppendSubimage mode. I could probably provide an example as well if you need it Larry.

**Comment by lgritz** [comment]
I can also reproduce the same problem from the command line as well:

```
$ oiiotool -pattern constant:color=1,0,0 64x64 3 -d float -o f.exr
$ oiiotool -pattern constant:color=0,1,0 64x64 3 -d half -o h.exr
$ oiiotool f.exr h.exr -siappend -o out.exr
$ iinfo -a -v out.exr
out.exr :   64 x   64, 3 channel, half openexr
    2 subimages: 64x64 64x64
 subimage  0:   64 x   64, 3 channel, half openexr
    channel list: R, G, B
    ...
 subimage  1:   64 x   64, 3 channel, half openexr
    channel list: R, G, B
    ...
```

The appended image should have a float subimage and a half subimage, but it's being saved as all half.
It's either a big coincidence, or the same problem underneath.

**Comment by lgritz** [comment]
Wait, I don't know if that's the same problem. I realize it's very hard to tell oiiotool that you want different formats for different subimages. Put that on hold, let me look at the Python issue first.

**Comment by lgritz** [comment]
I can certainly reproduce some weirdness with your script. Looking into it.

**Comment by lgritz** [comment]
I see the problem. Stay tuned.

**Comment by lgritz** [comment]
Proposed fix in #1316. Can you give that a try (it's a patch against RB-1.6) and let us know if it seems to fix everything for you?

**Comment by marcomeyerVFX** [comment]
This works for all my test cases so far! I was confused first, since iv marks all subimages as "half", but i assume that's due to the overall exr being labeled "half". So that's a different issue. According to iinfo and exrheader everything is nicely mixed now. Thanks a lot, Larry!!

**Comment by redesigndavid** [comment]
Will this fix also work on 1.5.* oiio versions?

**Comment by lgritz** [comment]
If you're not yet unable to upgrade to 1.6, I can backport the fix to 1.5.

**Comment by lgritz** [comment]
I pushed the fix to RB-1.5, but not yet tagged.
I'm going to do a round of tags/releases tomorrow (first of the month).

**Comment by redesigndavid** [comment]
Thank you very much Igritz!
